PROTO_tests/tests/bitops_test.py

```diff
@@ -0,0 +1,219 @@
+        par_ans = par_ans ^ 1
+        # clz_ans will be max_bits - 201 for all indices since that's the first nonzero bit everywhere now
+        # ctz_ans is unchanged other than first position which previously had no set bits
+        ctz_ans[0] = 200
+        for max_bits in [201, 256]:
+            bi.max_bits = max_bits
+            clz_ans = ak.full_like(bi, max_bits - 201)
+
+            assert pop_ans.to_list() == bi.popcount().to_list()
+            assert par_ans.to_list() == bi.parity().to_list()
+            assert clz_ans.to_list() == bi.clz().to_list()
+            assert ctz_ans.to_list() == bi.ctz().to_list()
+
+        # test with lots of trailing zeros
+        bi = ak.bigint_from_uint_arrays(
+            [
+                ak.arange(size, dtype=ak.uint64),
+                ak.zeros(size, dtype=ak.uint64),
+                ak.zeros(size, dtype=ak.uint64),
+            ]
+        )
+
+        # popcount and parity just look at number of bits, so this is equivalent to arange(10)
+        pop_ans = ak.popcount(i)
+        par_ans = ak.parity(i)
+        # ctz will include 2 new 64 bit arrays of zeros, but ctz(0) is still 0
+        ctz_ans = ak.ctz(i) + 128
+        ctz_ans[0] = 0
+
+        for max_bits in [138, 192, 201, 256]:
+            bi.max_bits = max_bits
+            # base_clz plus the amount that max_bits exceeds the bits used to store the bigint
+            clz_ans = base_clz + (max_bits - 192)
+            # except for the first position doesn't have any set bits, so we want the 128 bits after accounted
+            clz_ans[0] += 128
+
+            assert pop_ans.to_list() == bi.popcount().to_list()
+            assert par_ans.to_list() == bi.parity().to_list()
+            assert clz_ans.to_list() == bi.clz().to_list()
+            assert ctz_ans.to_list() == bi.ctz().to_list()
+
+        # test edge cases
+        edge_case = ak.cast(ak.array([-(2**63), -1, 2**63 - 1]), ak.uint64)
+        bi = ak.bigint_from_uint_arrays([edge_case, edge_case, edge_case])
+
+        pop_ans = ak.popcount(edge_case) * 3
+        # parity is the same as edge cases, because anything XORed with itself becomes zero so (x ^ x ^ x) = x
+        par_ans = ak.parity(edge_case)
+        base_clz = ak.clz(edge_case)
+        ctz_ans = ak.ctz(edge_case)
+
+        for max_bits in [192, 201, 256]:
+            bi.max_bits = max_bits
+            # base_clz plus the amount that max_bits exceeds the bits used to store the bigint
+            clz_ans = base_clz + (max_bits - 192)
+
+            assert pop_ans.to_list() == bi.popcount().to_list()
+            assert par_ans.to_list() == bi.parity().to_list()
+            assert clz_ans.to_list() == bi.clz().to_list()
+            assert ctz_ans.to_list() == bi.ctz().to_list()
+
+    @pytest.mark.parametrize("dtype", [bool, str, float])
+    def test_dtypes(self, dtype):
+        f = ak.cast(ak.array(range(10)), dtype)
+
+        with pytest.raises(TypeError):
+            ak.popcount(f)
```

Comment:
let's name this `arr` not `f` since that's usually only used for floats. Also we should continue to check both `arr.popcount()` and `ak.popcount(arr)`

I think renaming this would also be worthwhile since we're only testing error that the dtypes throw errors

Reply:
Updated